### PR TITLE
Баранов - Лабораторная работа #1

### PR DESCRIPTION
**Pull request by pavelbar**

### CODE REVIEW

File: code/string_calculator/src/string_calculator.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        self.string = string
+        self.delimiters = delimiters
+        self.numstring = numstring
+        self.numbers = numbers
+
+
+class StrCalculatorParser(object):
+    # string to StrCalculatorType converter
+    DEFAULT_DELIMS = ["\n", ","]
+    CUSTOM_DELIM_REGEX = "^([^\d])\n"
```
Comment (okainov) [inline code/string_calculator/src/string_calculator.py:19]: If the regexp is constant, why not precompile it before usage? Like `re.compile(...)`?

File: code/string_calculator/src/string_calculator.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    def parse(self, string):
+        m = self._check_custom_delim(string)
+        if m:
```
Comment (okainov) [inline code/string_calculator/src/string_calculator.py:23]: Can be shortened with `:=` operator with Python 3.8 =)

In any case, `m` is not the best variable name

File: code/string_calculator/src/string_calculator.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        integers = self._split_string(numstring)
+        return StrCalculatorType(string, delimiters, numstring, integers)
+
+    def _check_custom_delim(self, string):
+        return re.match(self.CUSTOM_DELIM_REGEX, string)
+
+    def _handle_custom_delim(self, string, m):
+        delimiters = m.captures(1)
+        numstring = re.sub(self.CUSTOM_DELIM_REGEX, "", string)
+        return (delimiters, numstring)
```
Comment (okainov) [inline code/string_calculator/src/string_calculator.py:37]: Extra parenthesis not needed

File: code/string_calculator/src/string_calculator.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _handle_default_delim(self, string):
+        delimiters = self.DEFAULT_DELIMS
+        numstring = string
+        return (delimiters, numstring)
+
+    def align_delimiters(self, delimiters, numstring):
+        for s in delimiters:
```
Comment (okainov) [inline code/string_calculator/src/string_calculator.py:45]: Iteration is not really efficient, why not use [`translate`](https://docs.python.org/3/library/stdtypes.html?highlight=maketrans#str.translate)?

File: code/string_calculator/src/string_calculator_test.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    def test_can_use_custom_delimiter_hard(self):
+        strcalc = StrCalculator()
+        result = strcalc.add(";\n1;2;3;4;5;6;7")
+        print(result)
+        self.assertTrue(result == 28)
+
+    def test_check_negative_numbers_exception(self):
+        strcalc = StrCalculator()
+        with self.assertRaises(Exception) as context:
```
Comment (okainov) [inline code/string_calculator/src/string_calculator_test.py:56]: - Exception too broad
- Why not use [`assertRaisesRegexp`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex)?

File: requirements.txt [hunk 6]
```diff
@@ -6,3 +6,4 @@ numpy==1.19.2
 pep8-naming==0.3.3
 pep8==1.7.1
 zipp==0.5.0
+regex==2020.10.28
```
Comment (okainov) [inline requirements.txt:9]: Any reason to choose `regexp` over built-in `re`? =)

Comment (pavelbar) [reply requirements.txt]: Done.

File: code/string_calculator/src/string_calculator_test.py (hunk 5 above)
Comment (pavelbar) [reply code/string_calculator/src/string_calculator_test.py]: Done.

File: code/string_calculator/src/string_calculator.py (hunk 4 above)
Comment (pavelbar) [reply code/string_calculator/src/string_calculator.py]: Done.

File: code/string_calculator/src/string_calculator.py (hunk 3 above)
Comment (pavelbar) [reply code/string_calculator/src/string_calculator.py]: Done.

File: code/string_calculator/src/string_calculator.py (hunk 2 above)
Comment (pavelbar) [reply code/string_calculator/src/string_calculator.py]: Done.

File: code/string_calculator/src/string_calculator.py (hunk 1 above)
Comment (pavelbar) [reply code/string_calculator/src/string_calculator.py]: Done.

File: code/string_calculator/src/string_calculator_test.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+class TestStrCalculatorClass(unittest.TestCase):
+    def test_can_create_calculator(self):
+        strcalc = StrCalculator()
+        self.assertTrue(isinstance(strcalc, StrCalculator))
+
+    def test_add_empty_string_is_0(self):
+        strcalc = StrCalculator()
+        result = strcalc.add("")
+        self.assertTrue(result == 0)
```
Comment (okainov) [inline code/string_calculator/src/string_calculator_test.py:16]: Why not assertEqual?

**Review by okainov — APPROVED** [review]
:+1:

**Comment by adgluzdov** [comment]
👍

**Comment by nikitavaneev** [comment]
+1

**Comment by nikitavaneev** [comment]
:+1:

**Comment by adgluzdov** [comment]
👍

File: requirements.txt [hunk 8]
```diff
@@ -5,4 +5,4 @@ nose==1.3.7
 numpy==1.19.2
 pep8-naming==0.3.3
 pep8==1.7.1
-zipp==0.5.0
+zipp==0.5.0
```
Comment (AntonGrishin) [inline requirements.txt:8]: looks like line endings style change, can you please remove this file from PR?

**Review by AntonGrishin — CHANGES_REQUESTED** [review]

**Comment by pavelbar** [comment]
@AntonGrishin done.

**Review by AntonGrishin — APPROVED** [review]

**Pull request merged**